Add time-synced lyrics overlay for main window and mini player

ef-molina opened the pull request:
- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [ ] This implements/fixes issue # (no existing issue)

---

## Description

This PR adds **time-synced lyrics (.lrc) support** to IINA with a Spotify-style overlay that works in both the **Main Window** and **Mini Player**.

The feature is fully additive and does not change existing playback, mpv integration, or UI behavior when lyrics are disabled.

---

## Feature Overview

- Displays synced lyrics over:
  - Main window (video / album art)
  - Mini player (music mode)
- Live updates based on playback time
- Shared lyrics visibility toggle across:
  - Main window
  - Mini player
  - Quick Settings
- Automatically hides when:
  - No lyrics are available
  - Lyrics are toggled off
- Restores lyrics immediately when toggled back on
- Adapts visible line count dynamically based on window size

---

## Architecture & Data Flow

**High-level flow:**

PlayerCore (playback time)
- LyricsController.syncTime(_:)
- LyricsEngine (pure logic)
- LyricsOverlayState
- NotificationCenter
- MainWindowController / MiniPlayerWindowController
- LyricsOverlayView (UI only)


### Key design points

- **LyricsEngine**
  - Pure, deterministic logic
  - Binary search for current line
  - Computes surrounding lyric window
  - No UI or PlayerCore dependencies

- **LyricsController**
  - Single source of truth for lyrics state
  - Receives playback time from PlayerCore
  - Emits notification updates only when the current line changes
  - Separates lyrics visibility from lyrics lifecycle (engine is preserved when toggling visibility)

- **LyricsOverlayView**
  - UI-only component
  - No knowledge of playback or notifications
  - Handles layout, font scaling, opacity hierarchy, and animations

- **Window Controllers**
  - Observe lyrics notifications
  - Control visibility
  - Forward state to views
  - Inject dynamic visible-line count via provider closure (no UI ↔ engine coupling)

- **LRCParser**
  - Parses external `.lrc` files only
  - Supports multi-timestamp lines
  - Ignores malformed entries
  - Does not assume lyric source priority

---

## Bug Fixes Included

- Fixed lyrics not reappearing after toggling OFF → ON (engine is no longer destroyed on visibility changes)
- Ensured Mini Player uses the same dynamic visible-line calculation as the Main Window
- Correctly clears lyrics when switching to tracks without `.lrc` files

---

## What This PR Does *Not* Include

- Automatic lyric fetching (e.g. lrclib)
- Embedded lyric parsing
- Caching or persistence
- Preference UI for lyric behavior
- Scroll or animation effects beyond opacity

These are intentionally left out to keep the initial implementation focused and reviewable.

---

## Testing

Manually tested on macOS:

- Main window lyrics overlay
- Mini player lyrics overlay
- Toggle ON / OFF / ON
- Window resizing
- Track changes (with and without lyrics)
- Seek / pause / resume playback

---

## Related Discussion

I opened up a related feature request/discussion here:
https://github.com/iina/iina/issues/5857

This PR is not intended to fully resolve that issue, but implements a focused,
opt-in synced lyrics overlay based on `.lrc` files.

---

## Notes

This is my first PR to IINA — happy to split, squash, or adjust anything if preferred by maintainers.

ef-molina committed:
Add LyricsLine model for synced lyrics

ef-molina committed:
Add LRC parser for synced lyrics

ef-molina committed:
Add lyrics engine for time-based lyric lookup

ef-molina committed:
Add synced lyrics overlay for music mode (initial)

- Add LyricsEngine overlay state computation
- Introduce LyricsController for time-based lyric sync
- Add LyricsOverlayView for 3-line lyric rendering
- Mount lyrics overlay in mini player (music mode)
- Wire PlayerCore time updates to lyrics controller

Notes:
- Initial UI focuses on mini player music mode only
- Lyrics are rendered as overlay on album art
- Future work: background contrast, multiline wrapping, toggles

ef-molina committed:
Added lyrics visibility state and notification plumbing

ef-molina committed:
Add lyrics visibility toggle to quick settings UI

ef-molina committed:
Implement synced lyrics overlay UI across main window and mini player

- Add LyricsOverlayView for rendering previous/current/next lines
- Wire LyricsController to drive overlay updates
- Integrate overlay into MainWindowController and MiniPlayerWindowController
- Ensure overlay does not intercept input events
- Handle layout transitions between mini and normal window modes

ef-molina committed:
Fix lyrics overlay not clearing when switching to tracks without .lrc files

When switching from a track with lyrics (.lrc file) to a track without
lyrics, the previous track's lyrics remained visible on screen. This was
because the clear() method in LyricsController only cleared the internal
engine state but didn't notify the overlay view to clear the displayed
lyrics.

The fix adds an explicit overlay update with an empty state when clear()
is called, ensuring the view clears the displayed lyrics immediately when
a track without lyrics is loaded.

ef-molina committed:
Added changes to dynamically display more or less lyrics lines dependent on window size

ef-molina committed:
Add lyrics toggle button to mini player controls

- Add lyrics visibility button to mini player control bar
- Sync toggle state with global lyrics visibility
- Update mini player XIB layout to include new control

ef-molina committed:
Lyrics overlay: fill album art container and compute visible line count dynamically

ef-molina committed:
Lyrics engine: support dynamic visible line counts

ef-molina committed:
Lyrics controller: forward dynamic visible line count to engine

ef-molina committed:
Main window: integrate dynamic lyrics overlay updates

ef-molina committed:
Refactor LyricsOverlayView: simplify layout and metrics

ef-molina committed:
Refactor LyricsEngine: clarify search and window logic

ef-molina committed:
Refactor LyricsController: clarify state and notification flow

ef-molina committed:
Docs: clarify LRCParser assumptions and behavior

ef-molina committed:
Fix lyrics toggle: keep engine when hiding overlay

ef-molina committed:
Fix mini player lyrics visible line count injection

uiryuu commented:
Thanks for proposing this PR. I personally very like this idea and the lyrics on top of NSVisualEffectView is great. However, a few things should be addressed.

1. For the lyrics toggle (in the main window), we should implement this toggle as a toolbar button on the OSC. Please revert any changes in the quicksettings and put a button in the OSC's toolbar.
2. The toggle button in the miniplayer should appear only if the lyrics file is found (`.lrc` file with same filename).
3. Currently the lyrics view is not very responsible. It will only update when the next line of lyrics arrives. It is desirable to update the lyrics view when the window dimensions change.

ef-molina commented:
@uiryuu 
Thanks for the review. I’ll update the PR to:
1. remove the Quick Settings lyrics toggle and add lyrics as an OSC toolbar button instead,
2. hide the Mini Player lyrics toggle unless a matching .lrc file is available,
3. force-refresh the lyrics overlay on window resize so the visible line count/layout updates immediately instead of waiting for the next lyric timestamp.

ef-molina commented:
@uiryuu  @low-batt 
pulled, rebased and pushing the requested changes.
1. Removed quick settings lyrics toggle. Added to OCS: Should it be visible similar to mini player or just standard on/off and if lyrics are available then display them. . . 

Toggle on the toolbar: pressing the toggle triggers persistant on or off and shows lyrics if available.
<img width="804" height="501" alt="Screenshot 2026-04-30 at 22 15 04" src="https://github.com/user-attachments/assets/78933f05-c44b-4ce2-8e1b-56623eb20fef" />

---

2. toggle button only appears if an .lrc file with the same name is available.
If lyrics are available in the mini player.

<img width="412" height="484" alt="Screenshot 2026-04-30 at 22 06 18" src="https://github.com/user-attachments/assets/6dfd624f-ebdf-442b-8bd3-8069112fa340" />

If lyrics are not found.
<img width="412" height="484" alt="Screenshot 2026-04-30 at 22 06 34" src="https://github.com/user-attachments/assets/46691969-0499-4b72-97ad-dc28754a14d6" />

---

3. the current implementation is now responsive, and force updates, not just when a new lyric is ready.

https://github.com/user-attachments/assets/327aab2f-0485-4b53-9272-7a528ed92f73


---

Note: Could dynamically add more lines of lyric lines based off of the size of the window. 
Let me know what you think or want to change or experiment with.

ef-molina committed:
Move lyrics toggle to OSC and gate Mini Player button

- remove the lyrics toggle from Quick Settings
- add lyrics as an OSC toolbar button with note/slash state
- preserve the user's lyrics visibility preference across tracks
- hide the Mini Player lyrics button unless usable same-name .lrc lyrics are loaded
- refresh lyrics overlay layout on window resize
- scope Mini Player lyrics notifications to the active player

uiryuu commented:
I think it looks nice now. Please push the changes so that we can review them

ef-molina commented:
> I think it looks nice now. Please push the changes so that we can review them

Pushed. LMK if anything needs to be fixed.

uiryuu commented on iina/Preference.swift:
```diff
@@ -852,6 +854,7 @@ struct Preference {
       case .subTrack: return makeSymbol(["captions.bubble.fill"], "sub-track")
       case .screenshot: return makeSymbol(["camera.shutter.button"], "screenshot")
       case .plugins: return makeSymbol(["puzzlepiece.extension"], "plugin")
+      case .lyrics: return makeSymbol(["music.note"], "toggle-album-art")
```
I suggest that we use "quote.bubble.fill" or "quote.bubble" for this button. Besides, for the fallback image, we should use a different one to prevent confusion. Fallback images will be used for systems below macOS 14.0. I suggest that we use "tab_sub" instead of "toggle-album-art" here.

uiryuu replied:
<img width="99" height="67" alt="image" src="https://github.com/user-attachments/assets/fe32e5ed-3c68-4328-862d-6765d89c9fc0" />

In apple music they also used "quote.bubble.fill" for lyrics

ef-molina replied:
Done. Changed the lyrics toolbar icon to `quote.bubble.fill` and changed the fallback image to `tab_sub`.

ef-molina replied:
Removed `(experimental)` from lyrics `MARK` comments.

ef-molina replied:
I added lyrics overlay refreshes for enter/exit PiP and enter/exit Mini Player.
and added an overlay bounds-change re-render so PiP resizing updates immediately instead of waiting for the next lyric timestamp.

uiryuu commented on iina/Preference.swift:
```diff
@@ -956,7 +960,7 @@ struct Preference {
     .controlBarAutoHideTimeout: Float(2.5),
     .enableControlBarAutoHide: true,
-    .controlBarToolbarButtons: [ToolBarButton.plugins.rawValue, ToolBarButton.pip.rawValue, ToolBarButton.playlist.rawValue, ToolBarButton.settings.rawValue],
+    .controlBarToolbarButtons: [ToolBarButton.lyrics.rawValue, ToolBarButton.plugins.rawValue, ToolBarButton.pip.rawValue, ToolBarButton.playlist.rawValue, ToolBarButton.settings.rawValue],
```
I don't think we should add this button to the default toolbar, so please revert this line

ef-molina replied:
Done. Removed the lyrics button from the default OSC toolbar button list. It remains available through toolbar customization.

uiryuu commented on iina/MainWindowController.swift:
```diff
@@ -736,11 +789,31 @@ class MainWindowController: PlayerWindowController {
   }
 
+  private func updateLyricsToolbarButton(_ button: NSButton) {
+    let symbolName = player.info.isLyricsVisible ? "music.note" : "music.note.slash"
```
Oh now I understand you probably chose this because it has `.slash` variant. I like the idea to change the toolbar button based on the current state, but we will try to introduce some universal way to update all toolbar buttons. For now please revert the changes to update toolbar button.

uiryuu commented on iina/MiniPlayerWindowController.swift:
```diff
@@ -410,6 +462,39 @@ class MiniPlayerWindowController: PlayerWindowController, NSPopoverDelegate {
+          accessibilityDescription: "Lyrics"
+      )?.withSymbolConfiguration(config)
+  }
+
```
For this part, remove the code to change the icon; we only want the code to support detecting lyrics and show the icon.

ef-molina replied:
Done. Removed the Mini Player icon-changing code. The Mini Player still detects lyrics availability, hides/shows the button based on `hasLyrics`, and updates the button state/overlay visibility.

uiryuu commented on iina/MiniPlayerWindowController.swift:
```diff
@@ -174,6 +216,7 @@ class MiniPlayerWindowController: PlayerWindowController, NSPopoverDelegate {
   // MARK: - Window delegate: Open / Close
 
   func windowWillClose(_ notification: Notification) {
+    
```
Unnecessary empty line

ef-molina replied:
Done. Removed the unnecessary blank line.

uiryuu commented on iina.xcodeproj/project.pbxproj:
```diff
@@ -2430,6 +2453,7 @@
 		84A0BAA31D2FAEA000BC8DA1 /* Views */ = {
 			isa = PBXGroup;
 			children = (
+				50D56BEC2EF8FB9600A37C7D /* LyricsOverlayView.swift */,
```
Consider putting `LyricsOverlayView.swift` in `Lyrics/`, and putting `Lyrics/` in `iina/Views/Accessories`.

ef-molina replied:
Done. Moved `LyricsOverlayView.swift` into `iina/Lyrics/` and moved the `Lyrics` group under `Views > Accessories` in the Xcode project navigator.

uiryuu commented on iina.xcodeproj/project.pbxproj:
```diff
@@ -2733,6 +2758,7 @@
 		84EB1ED81D2F51D3004FA5A1 /* iina */ = {
 			isa = PBXGroup;
 			children = (
+				50D56BE32EF8E04900A37C7D /* Lyrics */,
```
See above

ef-molina replied:
Done as part of the same project organization change: the `Lyrics` group now sits under `Views > Accessories`.

uiryuu commented on iina/Base.lproj/MiniPlayerWindowController.xib:
```diff
@@ -200,6 +201,20 @@
                                     </button>
+                                    <button hidden="YES" toolTip="Toggle Lyrics" translatesAutoresizingMaskIntoConstraints="NO" id="bFb-AY-7p7">
+                                        <rect key="frame" x="23" y="12" width="14" height="20"/>
+                                        <buttonCell key="cell" type="square" bezelStyle="shadowlessSquare" image="music.note" catalog="system" imagePosition="only" alignment="center" refusesFirstResponder="YES" imageScaling="proportionallyUpOrDown" inset="2" id="vHj-ED-sbh">
```
Use `quote.bubble.fill` and don't alter the icon based on the state for now.

ef-molina replied:
Done. Updated the Mini Player XIB to use `quote.bubble.fill`, and the icon is no longer altered based on lyrics state.

uiryuu commented on iina/Lyrics/LyricsLine.swift:
```diff
@@ -0,0 +1,14 @@
+struct LyricsLine: Equatable {
+    let time: TimeInterval
+    let text: String
+}
```
I think we can put `LyricsLine` and `LyricsOverlayState` into `LyricsEngine`. We don't enforce 1 class/file

ef-molina replied:
Done. Moved `LyricsLine` and `LyricsOverlayState` into `LyricsEngine.swift`, removed the separate files, and cleaned up the Xcode project references.

uiryuu commented on iina/PlayerCore.swift:
```diff
@@ -2675,6 +2715,11 @@ class PlayerCore: NSObject {
+
+      // MARK: - Lyrics Sync (experimental)
+      if let time = info.videoPosition?.second {
+        lyricsController.syncTime(time)
```
Only sync time with `lyricsController` when the lyrics overlay is available and visible

ef-molina replied:
Done. `lyricsController.syncTime` is now only called when lyrics are available, lyrics are visible, and a playback time exists.

uiryuu reviewed (changes requested):
We should also re-layout the lyrics layer when:
- Enter/exit pip
- Enter/exit mini player

Remove "(experimental)" in the MARK statement.

uiryuu commented on iina/Lyrics/LyricsEngine.swift:
```diff
@@ -0,0 +1,139 @@
+      return nil
+    }
+    return lines[index + 1]
+  }
```
It seems these 2 functions are not used?

ef-molina replied:
Done. Removed the unused `previousLine(at:)` and `nextLine(at:)` helpers.

uiryuu commented on iina/Lyrics/LyricsEngine.swift:
```diff
@@ -0,0 +1,139 @@
+  func linesAroundCurrent(at time: TimeInterval, count: Int) -> [LyricsLine] {
+    guard !lines.isEmpty, count > 0 else { return [] }
+    
+    guard let current = currentLine(at: time),
```
The only caller of `linesAroundCurrent` is in `overlayState`, at which you already have `current = currentLine(at: Time)`, so here we should use the current line as a parameter instead recompute it in `linesAroundCurrent`.

ef-molina replied:
Done. `overlayState` now computes the current line/index once, and `linesAroundCurrent` accepts the already-computed current index instead of recomputing from time.

ef-molina committed:
Address synced lyrics PR review feedback

ef-molina commented:
@uiryuu Thanks for the feedback, I addressed your comments.
Let me know if any additional changes are needed.

uiryuu reviewed (approved):
The changes LGTM now. Please wait for other dev's reviews.

lhc70000 committed:
Merge branch 'develop' into feature/synced-lyrics-main-window
